gateway-server/src/main/java/org/apache/knox/gateway/services/security/impl/DefaultKeystoreService.java

```diff
@@ -71,8 +74,9 @@
   private static GatewayMessages LOG = MessagesFactory.get(GatewayMessages.class);
   private static GatewayResources RES = ResourcesFactory.get(GatewayResources.class);
 
+  //let's configure the cache with hard-coded attributes now; we can introduce new gateway configuration later on if needed
+  private final Cache<String, Map<String, String>> cache = Caffeine.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).maximumSize(1000).build();
```

Comment:
Is it possible to have entries evicted after some duration has passed without a read?

Reply:
I believe it is possible. That rings the bell that we may want to keep the line which adds back an entry into the cache when the credential was loaded from the keystore.
Is this you'd like to point out?

Reply:
So the following should be implemented:

1. add an entry: into keystore and cache
2. 10 minutes passed w/o reading that entry -> got evicted from cache
3. fetch it -> we load it from the keystore but not adding into the cache

The least should be changed to add it back to the cache, right?

Reply:
At least that much. I'm thinking that it will be more efficient to keep the most-recently used entries in the cache, and evict those which have not been referenced for some time. So, if we could evict those entries which have not been read after n minutes, I think that would be preferable.

Reply:
This could be done easily (see `expireAfterAccess` [here](https://github.com/ben-manes/caffeine/wiki/Eviction)). Let me submit a new patch soon.

Reply:
Done.